### ISSUE TITLE
[Question] Does deepspeed take the average or sum of gradients across devices?

### ISSUE THREAD

**Issue by imoneoi** [microsoft/DeepSpeed#4662]
Does deepspeed average or sum gradients across devices? Is there any option through which we can choose this behavior?

In my training setup, each GPU will have a different number of tokens, and I divide the loss on each device by the total number of tokens on all GPUs, so I want deepspeed to sum the gradients across devices.

**Comment by loadams** [comment]
Gradient averaging is explained in the docs [here](https://www.deepspeed.ai/getting-started/#training), does that answer your question?

**Comment by DevHaufior** [comment]
https://github.com/microsoft/DeepSpeed/blob/901d80701258f59fc951c273f4c488ca04b10867/deepspeed/runtime/zero/stage_1_and_2.py#L1030

**Comment by RezaYazdaniAminabadi** [comment]
> Does deepspeed average or sum gradients across devices? Is there any option through which we can choose this behavior?
> 
> In my training setup, each GPU will have a different number of tokens, and I divide the loss on each device by the total number of tokens on all GPUs, so I want deepspeed to sum the gradients across devices.

I think this part is solved by having the sequence-parallelism set in deepspeed,. As @DevHaufior has shown in the code , by dividing the total world size by `sequence_parallel_size` you will exactly get what you need which is summing over the sequence-parallel ranks. This was solved by a recent [PR](https://github.com/microsoft/DeepSpeed/commit/ec029e7625ae3e6fb58c87a0483ea2c618ac3d3d) for DeepSpeed-ZeRO gradient averaging.

**Comment by loadams** [comment]
Thanks @RezaYazdaniAminabadi - @imoneoi does that help answer your question?
